src/init/common.cpp

```diff
@@ -80,15 +80,21 @@ util::Result<void> SetLoggingLevel(const ArgsManager& args)
 util::Result<void> SetLoggingCategories(const ArgsManager& args)
 {
     if (args.IsArgSet("-debug")) {
-        // Special-case: if -debug=0/-nodebug is set, turn off debugging messages
         const std::vector<std::string> categories = args.GetArgs("-debug");
 
-        if (std::none_of(categories.begin(), categories.end(),
-            [](std::string cat){return cat == "0" || cat == "none";})) {
-            for (const auto& cat : categories) {
-                if (!LogInstance().EnableCategory(cat)) {
-                    return util::Error{strprintf(_("Unsupported logging category %s=%s."), "-debug", cat)};
-                }
+        // Special-case: Disregard any debugging categories appearing before -debug=0/none
+        auto last_negated = categories.rbegin();
+        for (; last_negated != categories.rend(); ++last_negated) {
+            if (*last_negated == "0" || *last_negated == "none") {
+                break;
+            }
+        }
+
+        auto range_to_process = (last_negated == categories.rend()) ? categories : std::ranges::subrange(last_negated.base(), categories.end());
+
+        for (const auto& cat : range_to_process) {
+            if (!LogInstance().EnableCategory(cat)) {
+                return util::Error{strprintf(_("Unsupported logging category %s=%s."), "-debug", cat)};
```

Comment:
mega-nit: I find this marginally nicer, but still not great:
```suggestion
        for (auto it = last_negated == categories.rend() ? categories.begin() : last_negated.base();
            it != categories.end(); ++it) {
            if (!LogInstance().EnableCategory(*it)) {
                return util::Error{strprintf(_("Unsupported logging category %s=%s."), "-debug", *it)};
```